### PR TITLE
Feature/add sanity checks to write

### PR DESCRIPTION
**Pull request by janlemcke**

### CODE REVIEW

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 1]
```diff
@@ -14,19 +14,48 @@ const int32_t MB = 1000000;
...
     data_to_write = std::vector<int32_t>(vector_element_count, 42);
+    control_sum = vector_element_count * uint64_t{42};
 
     if (creat("file.txt", O_WRONLY) < 1) {
       std::cout << "create error" << std::endl;
     }
     chmod("file.txt", S_IRWXU);  // enables owner to rwx file
   }
 
+  void SanityCheck(int32_t NUMBER_OF_BYTES){
```
Comment (phoeinx) [inline src/benchmark/file_io_write_micro_benchmark.cpp:33]: Naming Scheme for functions is lower case with words separated by an underscore.
(I think we have to conform to Googles naming for functions for google benchmark provided functions like `TearDown()` and `SetUp()`, but everything else should follow hyrise naming conventions.

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 2]
```diff
@@ -14,19 +14,48 @@ const int32_t MB = 1000000;
 
 class FileIOWriteMicroBenchmarkFixture : public MicroBenchmarkBasicFixture {
  public:
+  uint64_t control_sum = uint64_t{0};
+
   void SetUp(::benchmark::State& state) override {
     // TODO(phoeinx): Make setup/teardown global per file size to improve benchmark speed
     ssize_t BUFFER_SIZE_MB = state.range(0);
     // each int32_t contains four bytes
     int32_t vector_element_count = (BUFFER_SIZE_MB * MB) / 4;
```
Comment (phoeinx) [inline src/benchmark/file_io_write_micro_benchmark.cpp:23]: not part of your code - but why is this an `int32_t` and not an `uint32_t` ?

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 3]
```diff
@@ -14,19 +14,48 @@ const int32_t MB = 1000000;
...
+    int32_t fd;
+    if ((fd = open("file.txt", O_RDONLY)) < 0) {
+      std::cout << "open error " << errno << std::endl;
+    }
+    chmod("file.txt", S_IRWXU);  // enables owner to rwx file
+    std::vector<uint32_t> read_data;
+    read_data.resize(NUMBER_OF_BYTES / 4);
+
+    // Getting the mapping to memory.
+    off_t OFFSET = 0;
```
Comment (phoeinx) [inline src/benchmark/file_io_write_micro_benchmark.cpp:43]: could prbly be const?

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 4]
```diff
@@ -14,19 +14,48 @@ const int32_t MB = 1000000;
...
     if (creat("file.txt", O_WRONLY) < 1) {
       std::cout << "create error" << std::endl;
     }
     chmod("file.txt", S_IRWXU);  // enables owner to rwx file
   }
 
+  void SanityCheck(int32_t NUMBER_OF_BYTES){
+    int32_t fd;
+    if ((fd = open("file.txt", O_RDONLY)) < 0) {
+      std::cout << "open error " << errno << std::endl;
```
Comment (phoeinx) [inline src/benchmark/file_io_write_micro_benchmark.cpp:36]:
```suggestion
      std::cout << "open error " << std::strerror(errno) << std::endl;
```

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 5]
```diff
@@ -14,19 +14,48 @@ const int32_t MB = 1000000;
...
+    }
+    chmod("file.txt", S_IRWXU);  // enables owner to rwx file
+    std::vector<uint32_t> read_data;
+    read_data.resize(NUMBER_OF_BYTES / 4);
+
+    // Getting the mapping to memory.
+    off_t OFFSET = 0;
+
+    int32_t* map = reinterpret_cast<int32_t*>(mmap(NULL, NUMBER_OF_BYTES, PROT_READ, MAP_PRIVATE, fd, OFFSET));
```
Comment (phoeinx) [inline src/benchmark/file_io_write_micro_benchmark.cpp:45]: why is the map of type `int32_t` if read_data is of type `uint32_t`?

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 6]
```diff
@@ -14,19 +14,48 @@ const int32_t MB = 1000000;
...
+    if (map == MAP_FAILED) {
+      std::cout << "Mapping Failed. " << std::strerror(errno) << std::endl;
+    }
+
+    memcpy(std::data(read_data), map, NUMBER_OF_BYTES);
+    auto sum = std::accumulate(read_data.begin(), read_data.end(), uint64_t{0});
```
Comment (phoeinx) [inline src/benchmark/file_io_write_micro_benchmark.cpp:51]: so we limit each element in the `read_data` vector to be of size `int32_t` and expect the sum to be expressable in `uint64_t`. With our current setup that should always work and I guess if we ever get overflow problems, we will find out and be able to debug it.

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 7]
```diff
@@ -14,19 +14,48 @@ const int32_t MB = 1000000;
...
+    Assert(control_sum == sum, "Sanity check failed: Not the same result");
```
Comment (phoeinx) [inline src/benchmark/file_io_write_micro_benchmark.cpp:52]: Maybe add something like - Got: Expected:

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 8]
```diff
@@ -14,19 +14,48 @@ const int32_t MB = 1000000;
...
+    chmod("file.txt", S_IRWXU);  // enables owner to rwx file
+    std::vector<uint32_t> read_data;
+    read_data.resize(NUMBER_OF_BYTES / 4);
+
+    // Getting the mapping to memory.
+    off_t OFFSET = 0;
+
+    int32_t* map = reinterpret_cast<int32_t*>(mmap(NULL, NUMBER_OF_BYTES, PROT_READ, MAP_PRIVATE, fd, OFFSET));
+    if (map == MAP_FAILED) {
+      std::cout << "Mapping Failed. " << std::strerror(errno) << std::endl;
```
Comment (phoeinx) [inline src/benchmark/file_io_write_micro_benchmark.cpp:47]:
```suggestion
      std::cout << "Mapping for Sanity Check Failed. " << std::strerror(errno) << std::endl;
```

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 9]
```diff
@@ -14,19 +14,48 @@ const int32_t MB = 1000000;
...
+    }
+
+    memcpy(std::data(read_data), map, NUMBER_OF_BYTES);
+    auto sum = std::accumulate(read_data.begin(), read_data.end(), uint64_t{0});
+    Assert(control_sum == sum, "Sanity check failed: Not the same result");
+    // Remove memory mapping after job is done.
+    if (munmap(map, NUMBER_OF_BYTES) != 0) {
+      std::cout << "Unmapping failed." << std::endl;
```
Comment (phoeinx) [inline src/benchmark/file_io_write_micro_benchmark.cpp:55]:
```suggestion
      std::cout << "Unmapping for Sanity Check failed." << std::endl;
```

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 10]
```diff
@@ -14,19 +14,48 @@ const int32_t MB = 1000000;
...
 class FileIOWriteMicroBenchmarkFixture : public MicroBenchmarkBasicFixture {
  public:
+  uint64_t control_sum = uint64_t{0};
+
   void SetUp(::benchmark::State& state) override {
     // TODO(phoeinx): Make setup/teardown global per file size to improve benchmark speed
     ssize_t BUFFER_SIZE_MB = state.range(0);
     // each int32_t contains four bytes
     int32_t vector_element_count = (BUFFER_SIZE_MB * MB) / 4;
     data_to_write = std::vector<int32_t>(vector_element_count, 42);
```
Comment (phoeinx) [inline src/benchmark/file_io_write_micro_benchmark.cpp:24]: not part of your code, but perhaps we shouldn't write exactly the same element over and over? Not sure if the OS optimizes on pages with always the same element. (It for sure does for 0-pages.)

File: src/benchmark/file_io_write_micro_benchmark.cpp [hunk 11]
```diff
@@ -14,19 +14,48 @@ const int32_t MB = 1000000;
...
+    control_sum = vector_element_count * uint64_t{42};
```
Comment (phoeinx) [inline src/benchmark/file_io_write_micro_benchmark.cpp:25]: maybe create a variable for the hardcoded input we currently have, to enforce always calculating the `control_sum` with the value elements used in the `data_to_write` vector.

File: src/benchmark/file_io_write_micro_benchmark.cpp (hunk 6 above)
Comment (phoeinx) [reply src/benchmark/file_io_write_micro_benchmark.cpp:51]: The `read_data` elements should however also be `uint`s to avoid underflow errors

**Review by phoeinx — CHANGES_REQUESTED** [review]

File: src/benchmark/file_io_write_micro_benchmark.cpp (hunk 10 above)
Comment (rrcomtech) [reply src/benchmark/file_io_write_micro_benchmark.cpp]: Lets put this to a later point in time and first merge this PR to get the code in.

**Review by rrcomtech — APPROVED** [review]

**Review by tjjordan — APPROVED** [review]

**Pull request merged**